README.md
# Slaxer
SlaxWeb Framework CLI tool


CHANGELOG.md
# ChangeLog

Changes between versions.

## Current changes

* add abstract command class
* load command parameters as services/properties from the service provider

## v0.3

### v0.3.0

* removed component command

## v0.2

## v0.2.0

* change config file name for new framework version

## v0.1

### v0.1.0

* console app service provider
* install additional component command
* send application instance as first parameter at command instantiation


tests/_support/_generated/UnitTesterActions.php
<?php  //[STAMP] afe236e24211133b48645948576bce9f
namespace _generated;

// This class was automatically generated by build task
// You should not change it manually as it will be overwritten on next build
// @codingStandardsIgnoreFile

use Codeception\Module\Asserts;
use Helper\Unit;

trait UnitTesterActions
{
    /**
     * @return \Codeception\Scenario
     */
    abstract protected function getScenario();

    
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that two variables are equal.
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertEquals()
     */
    public function assertEquals($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertEquals', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that two variables are not equal
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotEquals()
     */
    public function assertNotEquals($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotEquals', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that two variables are same
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertSame()
     */
    public function assertSame($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertSame', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that two variables are not same
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotSame()
     */
    public function assertNotSame($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotSame', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that actual is greater than expected
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertGreaterThan()
     */
    public function assertGreaterThan($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertGreaterThan', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * @deprecated
     * @see \Codeception\Module\Asserts::assertGreaterThen()
     */
    public function assertGreaterThen($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertGreaterThen', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that actual is greater or equal than expected
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertGreaterThanOrEqual()
     */
    public function assertGreaterThanOrEqual($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertGreaterThanOrEqual', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * @deprecated
     * @see \Codeception\Module\Asserts::assertGreaterThenOrEqual()
     */
    public function assertGreaterThenOrEqual($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertGreaterThenOrEqual', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that actual is less than expected
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertLessThan()
     */
    public function assertLessThan($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertLessThan', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that actual is less or equal than expected
     *
     * @param        $expected
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertLessThanOrEqual()
     */
    public function assertLessThanOrEqual($expected, $actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertLessThanOrEqual', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that haystack contains needle
     *
     * @param        $needle
     * @param        $haystack
     * @param string $message
     * @see \Codeception\Module\Asserts::assertContains()
     */
    public function assertContains($needle, $haystack, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertContains', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that haystack doesn't contain needle.
     *
     * @param        $needle
     * @param        $haystack
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotContains()
     */
    public function assertNotContains($needle, $haystack, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotContains', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that string match with pattern
     *
     * @param string $pattern
     * @param string $string
     * @param string $message
     * @see \Codeception\Module\Asserts::assertRegExp()
     */
    public function assertRegExp($pattern, $string, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertRegExp', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that string not match with pattern
     *
     * @param string $pattern
     * @param string $string
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotRegExp()
     */
    public function assertNotRegExp($pattern, $string, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotRegExp', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that variable is empty.
     *
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertEmpty()
     */
    public function assertEmpty($actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertEmpty', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that variable is not empty.
     *
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotEmpty()
     */
    public function assertNotEmpty($actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotEmpty', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that variable is NULL
     *
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNull()
     */
    public function assertNull($actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNull', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that variable is not NULL
     *
     * @param        $actual
     * @param string $message
     * @see \Codeception\Module\Asserts::assertNotNull()
     */
    public function assertNotNull($actual, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertNotNull', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that condition is positive.
     *
     * @param        $condition
     * @param string $message
     * @see \Codeception\Module\Asserts::assertTrue()
     */
    public function assertTrue($condition, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertTrue', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks that condition is negative.
     *
     * @param        $condition
     * @param string $message
     * @see \Codeception\Module\Asserts::assertFalse()
     */
    public function assertFalse($condition, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertFalse', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks if file exists
     *
     * @param string $filename
     * @param string $message
     * @see \Codeception\Module\Asserts::assertFileExists()
     */
    public function assertFileExists($filename, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertFileExists', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Checks if file doesn't exist
     *
     * @param string $filename
     * @param string $message
     * @see \Codeception\Module\Asserts::assertFileNotExists()
     */
    public function assertFileNotExists($filename, $message = null) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('assertFileNotExists', func_get_args()));
    }

 
    /**
     * [!] Method is generated. Documentation taken from corresponding module.
     *
     * Fails the test with message.
     *
     * @param $message
     * @see \Codeception\Module\Asserts::fail()
     */
    public function fail($message) {
        return $this->getScenario()->runStep(new \Codeception\Step\Action('fail', func_get_args()));
    }
}


tests/_support/_generated/FunctionalTesterActions.php
<?php  //[STAMP] 72d7b03ef9d64e3a484fd8745044f950
namespace _generated;

// This class was automatically generated by build task
// You should not change it manually as it will be overwritten on next build
// @codingStandardsIgnoreFile

use Helper\Functional;

trait FunctionalTesterActions
{
    /**
     * @return \Codeception\Scenario
     */
    abstract protected function getScenario();

    
}


src/AbstractCommand.php
<?php
namespace SlaxWeb\Slaxer;

use Psr\Log\LoggerInterface as Logger;
use Symfony\Component\Console\Command\Command;

/**
 * Base Command
 *
 * Abstract Command to simplify and unify initialization of the command. All SlaxWeb
 * Framework commands should extend from this base command.
 *
 * @package   SlaxWeb\Bootstrap
 * @author    Tomaz Lovrec <tomaz.lovrec@gmail.com>
 * @copyright 2017 (c) SlaxWeb
 * @license   MIT <https://opensource.org/licenses/MIT>
 * @link      https://github.com/slaxweb
 * @version   0.4
 */
abstract class AbstractCommand extends Command
{
    /**
     * Logger instance
     *
     * @var \Psr\Log\LoggerInterface
     */
    protected $logger = null;

    /**
     * Constructor
     *
     * Set the logger instance to class properties.
     *
     * @param \Psr\Log\LoggerInterface $logger Logger instance
     */
    public function __construct(Logger $logger)
    {
        $this->logger = $logger;

        parent::__construct();

        $this->logger->info(
            "Custom command has been initialized and configured",
            ["class" => get_class($this), "name" => $this->getName()]
        );
    }
}


src/Service/Provider.php
<?php
/**
 * Slaxer Service Provider
 *
 * Initiate the Symfony Console Component, and expose it to the DIC as a
 * service.
 *
 * @package   SlaxWeb\Slaxer
 * @author    Tomaz Lovrec <tomaz.lovrec@gmail.com>
 * @copyright 2016 (c) Tomaz Lovrec
 * @license   MIT <https://opensource.org/licenses/MIT>
 * @link      https://github.com/slaxweb/
 * @version   0.3
 */
namespace SlaxWeb\Slaxer\Service;

use Pimple\Container;
use Symfony\Component\Console\Application as CLIApp;
use SlaxWeb\Bootstrap\Commands\Component\InstallCommand;

class Provider implements \Pimple\ServiceProviderInterface
{
    /**
     * Register provider
     *
     * Register is called by the container, when the provider gets registered.
     *
     * @param \Pimple\Container $app Dependency Injection Container
     * @return void
     */
    public function register(Container $app)
    {
        $app["slaxer.service"] = function (Container $app) {
            $cliApp = new CLIApp("Slaxer", "0.4.*-dev");

            if (isset($app["slaxerCommands"]) === false) {
                return $cliApp;
            }

            foreach ($app["slaxerCommands"] as $key => $value) {
                $params = [];
                if (is_int($key)) {
                    $command = $value;
                } else {
                    $command = $key;
                    if (is_string($value)) {
                        $value = [$value];
                    }
                    foreach ($value as $service) {
                        $params[] = $app[$service];
                    }
                }

                $cmd = new $command($app["logger.service"]("Slaxer"));
                if (method_exists($cmd, "init")) {
                    $cmd->init(...$params);
                }
                $cliApp->add($cmd);
            }

            return $cliApp;
        };
    }
}


src/Component/RemoveCommand.php
<?php
/**
 * Slaxer Remove Component Command
 *
 * Remove Component command contains functionality to remove the command from the
 * Framework.
 *
 * @package   SlaxWeb\Slaxer
 * @author    Tomaz Lovrec <tomaz.lovrec@gmail.com>
 * @copyright 2016 (c) Tomaz Lovrec
 * @license   MIT <https://opensource.org/licenses/MIT>
 * @link      https://github.com/slaxweb/
 * @version   0.1
 *
 * @todo: needs a complete rewrite in the future, structure of the code here is catastrophic! Author: slax0r
 */
namespace SlaxWeb\Slaxer\Component;

use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class RemoveCommand extends BaseCommand
{
    /**
     * Configure the command
     *
     * Prepare the command for inclussion into the CLI Application Slaxer.
     *
     * @return void
     */
    protected function configure()
    {
        $this->setName("component:remove")
            ->setDescription("Remove SlaxWeb Framework Component")
            ->addArgument(
                "name",
                InputArgument::REQUIRED,
                "Which component do you want to remove?"
            );
    }

    /**
     * Execute the command
     *
     * Check that the component exists on packagist. If no slash is found in the
     * name, component name is automatically prepended by 'slaxweb/', if the package
     * exists, it checks that the 'composer' command is found and then proceeds by
     * removing the package with 'composer' if the component is installed.
     *
     * @param \Symfony\Component\Console\Input\InputInterface $input Command Input Object
     * @param \Symfony\Component\Console\Output\OutputInterface $output Command Output Object
     * @return void
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $this->input = $input;
        $this->output = $output;

        $component = $this->finalizeComponent([
            "name"          =>  strtolower($this->input->getArgument("name")),
            "version"       =>  $this->input->getArgument("version") ?? "",
            "installFlags"  =>  ""
        ]);

        if ($this->checkComposer() === false) {
            return;
        }

        if ($this->isInstalled($component["name"]) === false) {
            return;
        }

        $this->output->writeln("<comment>Removing component</>");
        if ($this->remove($component["name"]) === false) {
            $this->output->writeln(
                "<error>ERROR: An error occured while removing component. Check the output above.</>"
            );
        }
        $this->output->writeln("<comment>Component {$component["name"]} removed.</>");
    }
}


src/InstallComponentCommand.php
<?php
/**
 * Slaxer Service Provider
 *
 * Initiate the Symfony Console Component, and expose it to the DIC as a
 * service.
 *
 * @package   SlaxWeb\Slaxer
 * @author    Tomaz Lovrec <tomaz.lovrec@gmail.com>
 * @copyright 2016 (c) Tomaz Lovrec
 * @license   MIT <https://opensource.org/licenses/MIT>
 * @link      https://github.com/slaxweb/
 * @version   0.1
 */
namespace SlaxWeb\Slaxer;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class InstallComponentCommand extends Command
{
    /**
     * Guzzle Client
     *
     * @var \GuzzleHttp\Client
     */
    protected $_client = null;

    /**
     * Packagist Base Url
     *
     * @var string
     */
    protected $_baseUrl = "https://packagist.org/packages/";

    /**
     * Construct Class
     *
     * Store the GuzzleHTTP Client object to the class property.
     *
     * @param \GuzzleHttp\Client $client Guzzle Client
     * @return void
     */
    public function __construct(\GuzzleHttp\Client $client)
    {
        $this->_client = $client;
    }

    /**
     * Configure the command
     *
     * Prepare the command for inclussion into the CLI Application Slaxer.
     *
     * @return void
     */
    protected function configure()
    {
        $this->setName("component:install")
            ->setDescription("Install SlaxWeb Framework Component")
            ->addArgument(
                "name",
                InputArgument::REQUIRED,
                "Which component do you want to install?"
            );
    }

    /**
     * Execute the command
     *
     * Check that the component exists on packagist. If no slash is found in the
     * name, component name is automatically prepended by 'slaxweb/', so that
     * SlaxWeb components are installed by default. If the package exists, it
     * checks that the 'composer' command is found and then proceeds by
     * installing the package with 'composer'.
     *
     * @param \Symfony\Component\Console\Input\InputInterface $input Command Input Object
     * @param \Symfony\Component\Console\Output\OutputInterface $output Command Output Object
     * @return void
     *
     * @todo Run PostInstall.php script after package has been installed.
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $component = strtolower($input->getArgument("name"));
        if (strpos($component, "/") === false) {
            $component = "slaxweb/{$component}";
        }

        $output->writeln(
            "<comment>Checking if component {$component} exists ...</>"
        );
        if ($this->_checkComponentExists($component) === false) {
            $output->writeln("<error>Component {$component} not found.</>");
            return;
        }
        $output->writeln(
            "<comment>OK</>"
        );

        $output->writeln(
            "<comment>Checking if composer exists ...</>"
        );
        if (($cmd = $this->_getComposer()) === "") {
            $output->writeln("<error>Composer not found.</>");
            return;
        }
        $output->writeln(
            "<comment>OK</>"
        );

        $output->writeln(
            "<comment>Trying to install component {$component} ...</>"
        );
        $exit = 0;
        system("{$cmd} require {$component}", $exit);
        if ($exit !== 0) {
            $output->writeln("<error>Composer did not exit as expected.</>");
            return;
        }
        $output->writeln(
            "<comment>OK</>"
        );

        $output->writeln(
            "<comment>Component {$component} installed successfully.</>"
        );
    }

    /**
     * Check Component Exists
     *
     * Try to find the component on packagist.
     *
     * @param string $component Component name to check for existance.
     * @return bool
     */
    protected function _checkComponentExists(string $component): bool
    {
        $response = $this->_client->request(
            "GET",
            "{$this->_baseUrl}{$component}",
            ["allow_redirects" => false]
        );
        return $response->getStatusCode() === 200;
    }

    /**
     * Get Composer Command
     *
     * Get the composer command. Returns an empty string if no composer found.
     *
     * @return string
     *
     * @todo Install composer locally if not found
     */
    protected function _getComposer(): string
    {
        ($cmd = trim(`which composer`)) || ($cmd = trim(`which composer.phar`));
        return $cmd;
    }
}


src/Component/BaseCommand.php
<?php
/**
 * Slaxer Component Base Command
 *
 * Base Command for the Component set of commands includes functionality that is
 * same and/or similar accross all different types of Component commands.
 *
 * @package   SlaxWeb\Slaxer
 * @author    Tomaz Lovrec <tomaz.lovrec@gmail.com>
 * @copyright 2016 (c) Tomaz Lovrec
 * @license   MIT <https://opensource.org/licenses/MIT>
 * @link      https://github.com/slaxweb/
 * @version   0.1
 *
 * @todo: introduce some abstraction, right now it's just too procedural
 * @todo: needs a complete rewrite in the future, structure of the code here is catastrophic! Author: slax0r
 */
namespace SlaxWeb\Slaxer\Component;

use SlaxWeb\Bootstrap\Application;
use Symfony\Component\Console\Command\Command;

abstract class BaseCommand extends Command
{
    /**
     * Input
     *
     * @var \Symfony\Component\Console\Input\InputInterface
     */
    protected $input = null;

    /**
     * Output
     *
     * @var \Symfony\Component\Console\Output\OutputInterface
     */
    protected $output = null;

    /**
     * Composer executable
     *
     * @var string
     */
    protected $composer = "";

    /**
     * SlaxWeb Framework Instance
     *
     * @var \SlaxWeb\Bootstrap\Application
     */
    protected $app = null;

    /**
     * Guzzle Client
     *
     * @var \GuzzleHttp\Client
     */
    protected $client = null;

    /**
     * Error string
     *
     * @var string
     */
    protected $error = "";

    /**
     * Packagist Base Url
     *
     * @var string
     */
    protected $baseUrl = "";

    /**
     * Logger
     *
     * @var \Psr\Log\LoggerInterface
     */
    protected $logger = null;

    /**
     * Providers mapping
     *
     * Configuration file mapping for providers and their key names
     *
     * @var array
     */
    protected $providersMap = [
        "app" =>  [
            "file"  =>  "provider.php",
            "key"   =>  "providerList"
        ],
        "commands"  =>  [
            "file"  =>  "provider.php",
            "key"   =>  "commandsList"
        ],
        "hooks"     =>  [
            "file"  =>  "provider.php",
            "key"   =>  "hooksList"
        ]
    ];

    /**
     * Component meta data
     *
     * @var string
     */
    protected $metaData = [];

    /**
     * Init Command
     *
     * Store the GuzzleHTTP Client object to the class property.
     *
     * @param \SlaxWeb\Bootstrap\Application $app Framework instance
     * @param \GuzzleHttp\Client $client Guzzle Client
     * @return void
     */
    public function init(Application $app, \GuzzleHttp\Client $client)
    {
        $this->app = $app;
        $this->client = $client;

        $this->baseUrl = $this->app["config.service"]["slaxer.baseUrl"];
        $this->logger = $this->app["logger.service"]("Slaxer");

        $this->logger->info("Command '" . get_class($this) . "' has been initialized");
    }

    /**
     * Get component info
     *
     * Obtain component info from configuration if it exists, and was not passed
     * in as command line arguments.
     *
     * @param array $component Component data
     * @return array
     */
    protected function getComponent(array $component): array
    {
        $this->logger->info("Obtaining component information");

        $config = $this->app["config.service"]["slaxer.componentSettings"][$component["name"]] ?? [];
        $defVer = $this->app["config.service"]["slaxer.defaultVersion"]
                ?? "dev-master";

        if (strpos($component["name"], "/") === false) {
            $component["name"] = "slaxweb/{$component["name"]}";
        }

        if ($component["version"] === "") {
            $component["version"] = $config["version"] ?? $defVer;
        }

        $component["installFlags"] = $config["installFlags"] ?? "";

        $this->logger->debug("Gathered component information", $component);

        return $component;
    }

    /**
     * Check Composer Command
     *
     * Set the composer command. Returns bool(false) if no composer found.
     *
     * @return bool
     *
     * @todo: Install composer locally if not found
     */
    protected function checkComposer(): bool
    {
        $this->output->writeln("<comment>Checking if composer exists ...</>");

        ($this->composer = trim(`which composer`)) || ($this->composer = trim(`which composer.phar`));
        if ($this->composer === "") {
            $this->logger->error("Composer not found. Make sure you have it installed, and is executable in your PATH");
            $this->output->writeln(
                "<error>ERROR: Composer not found. Make sure you have it installed, and is executable in your PATH</>"
            );
            return false;
        }

        $this->output->writeln("<comment>OK</>");
        return true;
    }

    /**
     * Check Component Exists
     *
     * Try to find the component on packagist.
     *
     * @param string $component Component name to check for existance.
     * @return bool
     */
    protected function componentExists(string $component): bool
    {
        $this->output->writeln("<comment>Checking if component {$component} exists ...</>");

        $response = $this->client->request(
            "GET",
            "{$this->baseUrl}{$component}",
            ["allow_redirects" => false]
        );
        if ($response->getStatusCode() !== 200) {
            $this->logger->error("Component {$component} not found.");
            $this->output->writeln("<error>ERROR: Component {$component} not found.</>");
            return false;
        }

        $this->logger->info("Component {$component} has been found on packagist service");
        $this->output->writeln("<comment>OK</>");
        return true;
    }

    /**
     * Check component installed
     *
     * Check if component directory exists. If it does not, consider it removed.
     *
     * @param string $component Name of the component
     * @return bool
     */
    protected function isInstalled(string $component): bool
    {
        $this->output->writeln("<comment>Checking if component {$component} exists ...</>");

        if (file_exists("{$this->app["appDir"]}../vendor/{$component}/") === false) {
            $this->output->writeln(
                "<error>ERROR: Component directory not found, component '{$component}' not installed.</>"
            );
            return false;
        }

        $this->output->writeln("<comment>OK</>");
        return true;
    }

    /**
     * Parse component meta data
     *
     * Load the meta data of the component and parse it. If the meta data file is
     * not found, an error is set, and bool(false) is returned.
     *
     * @param string $name Name of the component
     * @return bool
     */
    protected function parseMetaData(string $name): bool
    {
        $metaFile = "{$this->app["appDir"]}../vendor/{$name}/component.json";
        if (file_exists($metaFile) === false) {
            system("{$this->composer} remove {$name}", $exit);
            $this->logger->error("Not a valid component. 'component.json' meta data file is missing. Package removed.");
            $this->error = "Not a valid component. 'component.json' meta data file is missing. Package removed.";
            return false;
        }

        $this->metaData = json_decode(file_get_contents($metaFile));
        $this->logger->debug("Component meta data parsed.", ["metaData" => $this->metaData]);
        return true;
    }

    /**
     * Remove component
     *
     * Removes the component, its config files, and set providers with the help
     * of composer.
     *
     * @param string $name Name of the component
     * @return bool
     */
    protected function remove(string $name): bool
    {
        $this->logger->info("Preparing to remove component {$name}");

        if ($this->parseMetaData($component["name"]) === false) {
            return false;
        }

        // remove providers from configuration
        if (empty($this->metaData->providers) === false) {
            foreach ($this->providersMap as $providerName => $map) {
                if (empty($this->metaData->providers->{$providerName}) === false) {
                    $this->removeProviders($map, $this->metaData->providers->{$providerName});
                }
            }
        }

        $this->logger->debug("All component providers have been removed from configuration");

        // Remove configuration files from framework configuration directory
        foreach ($this->metaData->configFiles as $file) {
            unlink("{$this->app["appDir"]}Config/{$file}");
        }

        $this->logger->debug("All component configuration files have been removed");

        $exit = 0;
        system("{$this->composer} remove {$name}", $exit);
        if ($exit !== 0) {
            $this->logger->error("Error removing installed component with composer.");
            return false;
        }

        $this->logger->info("Component {$name} has been removed.");
        return true;
    }

    /**
     * Configure installed component
     *
     * Add providers, hooks, configuration files, and install sub-components if user
     * requests it.
     *
     * @param string $name Component name
     * @return bool
     */
    protected function configComponent(string $name): bool
    {
        $this->logger->info("Running configuration of component '{$name}'");

        // add providers to configuration
        if (empty($this->metaData->providers) === false) {
            foreach ($this->providersMap as $providerName => $map) {
                if (empty($this->metaData->providers->{$providerName}) === false) {
                    $this->addProviders($map, $this->metaData->providers->{$providerName});
                }
            }
        }

        $this->logger->debug("Added component providers to configuration", ["providers" => $this->metaData->providers]);

        // Add configuration files to framework configuration directory
        foreach ($this->metaData->configFiles as $file) {
            copy(
                "{$this->app["appDir"]}../vendor/{$name}/config/{$file}",
                "{$this->app["appDir"]}Config/{$file}"
            );
        }

        $this->logger->debug("Copied component configuration files", ["files" => $this->metaData->configFiles]);

        // run post configure script
        if (empty($this->metaData->scripts->postConfigure) === false) {
            require "{$this->app["appDir"]}../vendor/{$name}/scripts/{$this->metaData->scripts->postConfigure}";
            $this->logger->debug(
                "Post configuration script has been executed.",
                ["script" => $this->metaData->scripts->postConfigure]
            );
        }

        $this->logger->info("Component configuration complete");
        return true;
    }

    /**
     * Add providers to config
     *
     * Add providers to provided config file, and the provided configuration key
     * name.
     *
     * @param array $config Configuration for provider including the file name and
     *                      configuration key name
     * @param array $providers List of providers to be added to configuration
     * @return void
     */
    protected function addProviders(array $config, array $providers)
    {
        // load config file
        $configFile = "{$this->app["appDir"]}Config/{$config["file"]}";
        $appConfig = file_get_contents($configFile);

        // get current providerList body
        preg_match("~\[[\"']{$config["key"]}['\"]\].+?\[(.*?)\];~s", $appConfig, $matches);
        $providerList = $matches[1];

        // append comma to last provider in list if needed
        preg_match_all("~^\s*?(['\"\\\\:\w\d_]+)(,?).*~m", $providerList, $matches);
        if (end($matches[2]) === "") {
            $newList = str_replace(end($matches[1]), end($matches[1]). ",", $providerList);
        } else {
            $newList = $providerList;
        }

        foreach ($providers as $provider) {
            if (strpos($newList, $provider) === false) {
                $newList .= "\n{$provider}::class,";
            }
        }
        $newList = rtrim($newList, ",") . "\n";

        $appConfig = str_replace($providerList, $newList, $appConfig);

        file_put_contents($configFile, $appConfig);
    }

    /**
     * Remove providers from config
     *
     * Remove providers from the provided config file, and the provided configuration
     * key name.
     *
     * @param array $config Configuration for provider including the file name and
     *                      configuration key name
     * @param array $providers List of providers to be added to configuration
     * @return void
     */
    protected function removeProviders(array $config, array $providers)
    {
        // load config file
        $configFile = "{$this->app["appDir"]}Config/{$config["file"]}";
        $appConfig = file_get_contents($configFile);

        foreach ($providers as $provider) {
            // remove each provider from the list
           $appConfig =  preg_replace(
                "~(\[[\"']{$config["key"]}['\"]\].+?\[.*){$provider}::class,?\n?(.*\];)~s",
                "$1$2",
                $appConfig
            );
        }

        // rewrite config file
        file_put_contents($configFile, $appConfig);
    }
}
